Source set configuration for plugin properties

AdamBJ opened the pull request:
This PR does two things (and consequently probably should be two PRs instead of one, but oh well.)

1. Shifts the plugin configuration information from the `project` object to the `generatedSource` source set (see commit c8dbc3b and #8). Also changes the integration test to verify the new behavior (see commit f9c901b).
2. Changes where `TestProject`, the Gradle project used for the integration tests, looks for the Umple plugin jar. Rather than looking to plugin portal, it now uses the locally built version of the plugin jar. Doing it this way makes it easier to test changes to the plugin before they are pushed to the plugin portal. A new task `copyPluginJars` was added to the Umple Gradle plugin's `build.gradle` file to support this.

closes #8

AdamBJ committed:
Changing where the plugin configuration information is stored and read from.

- UmpleGradlePlugin.groovy:
 - Enable to plugin to read configuration information from the generatedSource source set.
 - Minor refactoring.
- README.md
 - Updating to reflect the changes to UmpleGradlePlugin.groovy

AdamBJ committed:
Tweaking integrationTest/groovy/CompilationAndGeneration.groovy to test the sourceSet property changes in the previous commit. More work could be done here, but I think it's good enough for now.

AdamBJ committed:
Changing TestProject/build.gradle so that TestProject now gets the plugin jar it uses from the local machine instead of the plugin portal. This allows developers to more easily test their changes before publishing a new version of the plugin.

Nava2 commented on README.md:
````diff
@@ -1,26 +1,32 @@
-project.ext.set("umpleFileName", "test.ump")
-project.ext.set("languageToGenerate", "Java")
-```
+These values are set in the build.gradle file of the project you're applyung the plugin to. If any required fields are not set, the plugin will use the following default values:
````
typo: `applyung`

Nava2 commented on README.md:
```diff
@@ -1,26 +1,32 @@
-gradle compileUmpleFile -PumpleFileName=test.ump -PlanguageToGenerate=Java
+sourceSets {
+    generatedSource { 
+       ext.languageToGenerate = 'Java' 
```
Do you need to use the `ext` here? I thought the link you sent previously wasn't necessary. 

Could we create a new `sourceSet` alternately?

AdamBJ replied:
>Do you need to use the ext here? I thought the link you sent previously wasn't necessary.
In the example they actually declare the `purpose` property first: 

`sourceSets.all { ext.purpose = null }`

There's no way around using the `ext` as far as I can tell.

>Could we create a new sourceSet alternately?

Not at the moment. If they don't declare a `generatedSource` source set Gradle throws an exception. Allowing the user to customize the name of the source set is tricky. We would need to pass the source set name to the plugin, but we currently rely on the `generatedSource` source set to pass all our information for us. There's a chicken-or-the-egg problem in there somewhere.

AdamBJ replied:
@Nava2 One thing I don't like about the source set approach is that it makes generating non-java files a bit awkward. If the user wants to generate something other than Java they still need to apply the java plugin and create a `generatedSource` source set. It doesn't feel very ergonomic to be have to do so much java stuff if you're not generating Java files.

Using the `project` to pass configuration information had its drawbacks, but it had a lot less boilerplate.

AdamBJ replied:
I think using source sets is still a good idea. For one thing, it allows for use of the `compileGeneratedSourceJava` task if the generated files are Java. I just wish there was a way to make it a bit more light-weight.

Nava2 replied:
I've been playing around with this.. I have some ideas I'll get back to you tomorrow. Though, I think this comment thread is okay for this PR. :)

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -18,8 +18,8 @@ public class CompilationAndGeneration {
         builder.put(DIRECTORIES_ROOT + "sub2/build/classes/generatedSource", ["Sub2.class", "Sub22.class"])
-        builder.put(DIRECTORIES_ROOT + "subproj/generated/java", ["Subproj.java", "Subproj2.java"])
-        builder.put(DIRECTORIES_ROOT + "sub2/generated/java", ["Sub2.java", "Sub22.java"])
+        builder.put(DIRECTORIES_ROOT + "subproj/customPath/generated/java", ["Subproj.java", "Subproj2.java"])
```
So is the "customPath" being used to test for changing the output? Could we create multiple tests to check this instead of wrapping them into one? For example, one test where it's default, one where it's custom?

AdamBJ replied:
I agree we should be testing both. I chose to test the custom fields here because I think those are higher priority.

Of the top of my head, the only way I can see to add another set of tests would be to add another test project. Gradle isn't very flexible in terms of what you can do to manipulate a project during testing. Adding a new project would add a lot of duplicate code and take time, and I really want to focus on getting the compiler jar built. Also, I'm sure there is a smarter way to add another set of tests. It'll just take a bit of thinking to figure it out.

What I'd like to do is update #7 so that it's clear that more tests need to be added. Then, if I have time I'll get to it after the umple jar. If not, then then next person can tackle it.

Nava2 replied:
Can you make a "checklist" of tests that _should_ be added but haven't yet? I think that's a fair compromise!

AdamBJ replied:
@Nava2 : https://github.com/umple/umple.gradle/issues/7#issuecomment-286629268

Nava2 replied:
:+1:

Nava2 commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -3,16 +3,11 @@ buildscript {
-    classpath "gradle.plugin.cruise.umple:UmpleGradlePlugin:0.1.2"
-    //classpath "umple.gradle.plugin:UmpleGradlePlugin-0.1.2"
+    // Use the locally built version of the plugin for the tests
+    classpath files('libs/UmpleGradlePlugin-0.1.3.jar')
```
Is there some way to get this version number from a config? Loading for both the actual build's number and the tests? Perhaps via some kind of include?

AdamBJ replied:
I'll look into that

AdamBJ replied:
@Nava2 This is proving surprisingly difficult. The trick is that the main `umple.gradle` project and `TestProject` are completely separate - nothing passes between them.  That rules out anything other than some kind of `Constants.java` file they could share. But I can't figure out how to work with static variables from another file in `build.gradle`. When I do something like `... = Constants.VERSION_NUMBER` Gradle thinks I'm referring to a `Constants` property that doesn't exist.

I posted a question on SO. Hopefully someone has an idea. I'll keep thinking too.

Nava2 replied:
I think you want settings https://docs.gradle.org/current/dsl/org.gradle.api.initialization.Settings.html

AdamBJ replied:
I'm still not sure about this. I can add `TestProject` as a child project of the root project (the plugin), but I can't set an `ext` field or anything like that from inside `settings.gradle`. In fact, the kinds of project settings I change change are quite limited: https://docs.gradle.org/current/javadoc/org/gradle/api/initialization/ProjectDescriptor.html

Pretty much just the name of the build file, the project name, etc can be changed. I don't see how I can add a version number that's accessible to both projects. I also don't like the idea of making `TestProject` a sub project of the gradle plugin project. I feel that a sub project should be directly involved in the build somehow. `TestProject` is only used in integration testing.

AdamBJ replied:
@Nava2 Are your changes going to affect the plugin version stuff too? I've got something to try that might allow us to use a `Constants.java` file in both build files.

Nava2 replied:
No, hadn't tried it. Though, I think adding build logic is the answer.

AdamBJ replied:
I agree that using build logic would be best. But if it came down to leaving it as-is or using a constants file, which do you think would be better?

Nava2 commented:
So, I'm working on your branch, and when I run the integration tests, I don't think the `generateSource` task is called. I added a `throw new IllegalStateException()` into the task and it never errors. 

Edit: Are the integration tests running against the current sources or the hosted one?

AdamBJ commented:
😱 

I'm looking into this now. On the one hand, I don't know how it's possible. The test deletes all the generated files after it runs, so if the generate task isn't running the tests should be failing. No generated files are uploaded either, so I don't know how Travis is passing if the generate task isn't running. @Nava2 Are the integration tests failing on your end? 

On the other hand, having TestProject run `compileGeneratedSourceJava` in the integration test rather than `compileJava` seems to fix the issue (i.e. it causes the generate task to run properly).

The integration tests are running against the current source.

AdamBJ commented:
I think the big question here is whether `compileJava` compiles all the source sets or not. I thought that all source sets would automatically `dependsOn` the `compileJava` task.

AdamBJ commented:
@Nava2 are the tests still passing on your  machine despite the generateSource task not serving to work? I wonder if the issue is that the plugin jar isn't being updated and copied  properly.

AdamBJ commented:
@Maybe the problem is that the jar with the illegal state exception included isn't being used by `Test Project`

Nava2 commented:
> @Nava2 are the tests still passing on your machine despite the generateSource task not serving to work? I wonder if the issue is that the plugin jar isn't being updated and copied properly.

Yup! The modified version wasn't changing.

AdamBJ commented:
That's very weird. I'm hoping it's a jar copy issue. If the `generateSource` task isn't running, the tests should fail, period. Especially the tests for `java` files since there's the additional content verification step for those. I've got a couple things on my plate right now, but I'm going to dive into this issue early next week.

AdamBJ commented:
@Nava2 I think I've figured out what's going on. My guess is that after making the changes to `UmpleGradlePlugin.groovy` that add the `IllegalStateException` you directly ran the `integrationTest` task? If you did it that way, the problem occurred because the `integrationTest` task doesn't rebuild the plugin jar. It just copies whatever is in the `src/integrationTest/resources/TestProject/libs` to a location that `TestProject` can access. 

To regenerate the jar the `build` task needs to be run. My guess is that you didn't run the `build` task to rebuild the jar before running `integrationTest`, so what was copied over by `copyPluginJars` was an older version of the jar without the exception changes. That's what I'm hoping anyway!

Can you try running `gradle build` and let me know if you encounter the `IllegalStateException`? The generate task is definitely running on my end -- I commented out the `destroy` method in the integration tests and examined the generated java files. They're all there! Whether or not the dependencies of the `integrationTest` task need to change is another matter, but let's make sure we understand the cause of the problem before we make a decision on that front.

Nava2 commented:
Could we add better dependencies to fix this issue? I'll try tomorrow.

AdamBJ commented:
We could add the `jar` task as a dependency of the `integrationTest` task. That way whenever the integration tests are run we'd be using the most up-to-date version of the plugin. 

That's a one-line change: `integrationTest.dependsOn('copyUmpleJar', 'copyPluginJars', 'jar')`

Nava2 commented:
> We could add the jar task as a dependency of the integrationTest task. That way whenever the integration tests are run we'd be using the most up-to-date version of the plugin.
> 
> That's a one-line change: integrationTest.dependsOn('copyUmpleJar', 'copyPluginJars', 'jar')

I think this makes more sense: `copyPluginJars.dependsOn('jar')` but definitely the right idea.

AdamBJ committed:
Addresses some of Kevin's comments on PR#13, major README.md revision, and minor error-handling improvements

- Fixes typos in README.md
- Reworks README.md to better reflect the changes that have been made to the plugin over the semester
- If the plugin can't find the `generatedSource` source set it will now throw a `GradleException` with an informative error message
- If we can't find a folder in `CompilationAndGeneration.groovy` the tests now exit immediately with a more informative error message rather than continuing to run.

AdamBJ commented:
I've pushed some changes that deal with the `integrationTest` issue and fix the typo in `README.md` you found. The push also includes some better error-handling and a major reworking of `README.md`. 

What is doesn't address is your question about removing the hard-coded version numbers in the `TestProject` and `umple.gradle` `build.gradle` files (see my comments on that). It also doesn't do anything about the use of `ext` you mentioned. Let me know when you've had a chance to try out your idea to help us move away from `ext`.

Nava2 commented:
So I've essentially rewritten a good chunk of things to model after the `Scala` and `Antlr` plugins.. I'll have a branch in a bit. 😈

AdamBJ commented:
Oh boy 😨

Nava2 commented:
@AdamBJ have you been able to debug the plugin itself or only the test code?

AdamBJ commented:
@Nava2 I'm not sure what you mean. I'm able to run the plugin's `build` task or `integrationTest` task to run the integration tests, which involves the `TestProject`. I can also treat the `TestProject` as an dependent project and generate test files that way. Either way, the files are generating and the tests are passing.

Nava2 commented:
I mean, like stepping into the code itself.

AdamBJ commented:
I haven't tried using a debugger. I just ran the task and verified that the outputs were as expected.

Nava2 commented:
@AdamBJ can you please look at the comments I made in https://github.com/umple/umple.gradle/tree/ss-config-n2

Nava2 commented:
This is the base idea: https://github.com/gradle/gradle/tree/master/subprojects/scala
